File: session/01_concurrency_and_parallelism/parallelism/java/Portfolio.java
import java.util.*;
import java.util.stream.*;
import java.util.concurrent.*;

public class Portfolio {
  private Map<String, Integer> stocks = new HashMap<>();
  
  public void add(String ticker, Integer qty) {
    int oldQty = 0;
    if (stocks.containsKey(ticker)) {
      oldQty = stocks.get(ticker);
    }
    stocks.put(ticker, oldQty + qty);
  }
	
  public Double netWorth(NationalStockService stockService) throws Exception {
    System.out.println("Stocks = " + stocks);
    List<Double> prices = stocks.entrySet()
			.stream()
			.collect(ArrayList<Double>::new, (acc, entry) -> {
        String ticker = entry.getKey();
        try {
          acc.add(stockService.getPrice(ticker) * entry.getValue());  
        } catch (Exception e) {
          e.printStackTrace();
        }
      }, ArrayList::addAll);
    double worth = prices.stream().reduce(0d, (a, e) -> a + e);
    return worth;
  }
  
  public static void main(String[] args) throws Exception {
    Thread.sleep(10000);
    Portfolio portfolio = new Portfolio();
    portfolio.add("GOOG", 10);
    portfolio.add("AAPL", 20);
    portfolio.add("YHOO", 30);
    portfolio.add("MSFT", 40);
    System.out.println("NetWorth = " + portfolio.netWorth(new NationalStockService()));
    System.out.println("DONE");
  }
}

File: melodies/concurrency_and_parallelism/parallelism/splitting_io_task/scala/Portfolio.scala
package tsys.stocks

case class Portfolio() {
  // private val stocks = new scala.collection.mutable.HashMap[String, Int]()
  //
  // def add(ticker: String ticker, qty: Int qty) = {
  //   val oldQty = 0
  //   if (stocks(ticker)) {
  //     oldQty = stocks(ticker)
  //   }
  //   stocks += (ticker -> oldQty + qty)
  // }
  //
  // def netWorth(stockService: NationalStockService): Double = {
  //   // ForkJoinPool pool = new ForkJoinPool(10);
  //   val startTime = System.currentTimeMillis()
  //   stocks.keys
  //     // comment or uncomment the parallel() call below and see the difference
  //     .par
  //       .foldLeft(0d) { case(ticker, qty) =>
  //       acc.add(stockService.getPrice(ticker) * entry.getValue());
  //       }
  //         String ticker = entry.getKey();
  //         try {
  //
  //         } catch (Exception e) {
  //           e.printStackTrace();
  //         }
  //       }, ArrayList::addAll);
  //       val worth = prices.stream().reduce(0d, (a, e) -> a + e);
  //       val timeTaken = System.currentTimeMillis() - startTime
  //       println(String.format("Overall Time %d(ms)", timeTaken));
  //       worth
  //   // pool.shutdown();
  //   // return worth.get();
  // }
  
  private val stockService = new NationalStockService

  def track(tickers: String*): List[Double] = {
    tickers.par.flatMap(ticker => {
      val price = stockService.getPrice(ticker).toOption.toList
      println(s"Got $ticker $price using ${Thread.currentThread()}")
      price
    }).toList
    // val prices = new scala.collection.mutable.MutableList[Double]()
    // // tickers.foreach(ticker => stockService.getPrice(ticker).foreach(price => prices += price))
    // tickers.par.foreach(ticker => stockService.getPrice(ticker).foreach(price => {
    //   println(s"Got $ticker $price using ${Thread.currentThread()}")
    //   prices += price
    // }))
    // prices.toList
  }
}
